maykin_common/pdf.py

```diff
@@ -57,84 +111,69 @@ class UrlFetcherResult(TypedDict):
 
 class UrlFetcher:
     """
-    URL fetcher that skips the network for /static/* files.
+    URL fetcher that skips the network for /static/* and /media/* files.
     """
 
-    def __init__(self):
-        self.static_url = self._get_fully_qualified_url(settings.STATIC_URL)
-        is_static_local_storage = issubclass(
-            staticfiles_storage.__class__, FileSystemStorage
-        )
-
-        self.media_url = self._get_fully_qualified_url(settings.MEDIA_URL)
-        is_media_local_storage = issubclass(
-            default_storage.__class__, FileSystemStorage
-        )
-
-        self.candidates = (
-            (self.static_url, staticfiles_storage, is_static_local_storage),
-            (self.media_url, default_storage, is_media_local_storage),
-        )
-
-    @staticmethod
-    def _get_fully_qualified_url(setting: str):
-        fully_qualified_url = setting
-        if not urlparse(setting).netloc:
-            fully_qualified_url = urljoin(get_base_url(), setting)
-        return urlparse(fully_qualified_url)
-
     def __call__(self, url: str) -> UrlFetcherResult:
-        orig_url = url
+        """
+        Check if the URL matches one of our candidates and use it if there's a match.
+
+        Matching is done on the URLs of the storages and the requested asset. If the
+        prefix matches, look up the relative asset path in the storage and serve it
+        if it's found. If not, defer to the default URL fetcher of WeasyPrint.
+        """
         parsed_url = urlparse(url)
+        assert parsed_url.netloc and parsed_url.netloc, "Expected fully qualified URL"
 
-        candidate = self.get_match_candidate(parsed_url)
-        if candidate is not None:
-            base_url, storage = candidate
-            path = PurePosixPath(parsed_url.path).relative_to(base_url.path)
+        # Try candidates, respecting the order of the candidate configuration.
+        for base, storage in _get_candidate_storages().items():
+            base_url = base.geturl()
+            # Skip to the next candidate if the URLs don't share a prefix.
+            if not url.startswith(base_url):
```

Comment:
Is there some normalisation going on? Should we consider:
```suggestion
            if not parsed_url.geturl().startswith(base_url):
```

I don't think so. Unsure.

Reply:
I had the same thoughts. Given a URL with trailing `?`, it gets normalized to a string without the `?` because there are no query params at all, but otherwise I think it's not relevant because in all realistic cases you configure static url to be `[{scheme}{netloc}]{path}` which always acts as a prefix. So whatever may be trailing after the base url is irrelevant anyway. The check before was checking that scheme and netloc match, and then prefix match on the path.

I picked the less complex option because of this context.